File: frontend/src/api/renderer.ts
import { apiClient, ApiResponse } from './client';

export type RendererKind = 'builtin' | 'api' | 'component';
export type RendererScope = 'global' | 'user';

export interface RendererInfo {
  id: string;
  name: string;
  description: string;
  kind: RendererKind;
  extensions: string[];
  scope: RendererScope;
  user_id?: string | null;
  enabled: boolean;
  api_url?: string | null;
  entry?: string | null;
  config?: Record<string, any>;
  owner_user_id?: string | null;
  installed: boolean;
  source: 'official' | 'personal';
}

export interface RendererRequest {
  name: string;
  description: string;
  kind: RendererKind;
  extensions: string[];
  enabled: boolean;
  api_url?: string | null;
  entry?: string | null;
  config?: Record<string, any>;
  is_global: boolean;
}

interface RendererListResponse {
  renderers: RendererInfo[];
}

interface RendererUpsertResponse {
  renderer: RendererInfo;
}

export async function listRendererConfigs(): Promise<RendererInfo[]> {
  const response = await apiClient.get<ApiResponse<RendererListResponse>>('/renderers');
  return response.data.data.renderers;
}

export async function listRendererCatalog(): Promise<RendererInfo[]> {
  const response = await apiClient.get<ApiResponse<RendererListResponse>>('/renderers/catalog');
  return response.data.data.renderers;
}

export async function installRenderer(id: string): Promise<RendererInfo> {
  const response = await apiClient.post<ApiResponse<RendererInfo>>(`/renderers/${encodeURIComponent(id)}/install`);
  return response.data.data;
}

export async function uninstallRenderer(id: string): Promise<RendererInfo> {
  const response = await apiClient.delete<ApiResponse<RendererInfo>>(`/renderers/${encodeURIComponent(id)}/install`);
  return response.data.data;
}

export async function createRendererConfig(request: RendererRequest): Promise<RendererInfo> {
  const response = await apiClient.post<ApiResponse<RendererUpsertResponse>>('/renderers', request);
  return response.data.data.renderer;
}

export async function updateRendererConfig(id: string, request: RendererRequest): Promise<RendererInfo> {
  const response = await apiClient.put<ApiResponse<RendererUpsertResponse>>(`/renderers/${encodeURIComponent(id)}`, request);
  return response.data.data.renderer;
}

export async function deleteRendererConfig(id: string): Promise<RendererInfo[]> {
  const response = await apiClient.delete<ApiResponse<RendererListResponse>>(`/renderers/${encodeURIComponent(id)}`);
  return response.data.data.renderers;
}

File: backend/app/interfaces/schemas/renderer.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.domain.models.renderer import RendererKind, RendererScope


class RendererResponse(BaseModel):
    id: str
    name: str
    description: str = ""
    kind: RendererKind
    extensions: List[str] = Field(default_factory=list)
    scope: RendererScope
    user_id: Optional[str] = None
    owner_user_id: Optional[str] = None
    workspace_id: Optional[str] = None
    enabled: bool = True
    api_url: Optional[str] = None
    entry: Optional[str] = None
    config: Dict[str, Any] = Field(default_factory=dict)
    installed: bool = False
    source: str = "personal"


class RendererListResponse(BaseModel):
    renderers: List[RendererResponse] = Field(default_factory=list)


class RendererRequest(BaseModel):
    name: str
    description: str = ""
    kind: RendererKind = RendererKind.API
    extensions: List[str] = Field(default_factory=list)
    enabled: bool = True
    api_url: Optional[str] = None
    entry: Optional[str] = None
    config: Dict[str, Any] = Field(default_factory=dict)
    is_global: bool = False


class RendererUpsertResponse(BaseModel):
    renderer: RendererResponse
